File: packages/sdk-ruby/lib/melaya/evals.rb
# frozen_string_literal: true

module Melaya
  # Evals API — list and inspect agent evaluation run results.
  #
  # Maps to /api/v1/private/evals/*.
  #
  # @example
  #   runs    = melaya.evals.list_runs
  #   summary = melaya.evals.summary
  #   detail  = melaya.evals.run_detail(runs.first["id"])
  class EvalsAPI
    def initialize(http)
      @http = http
    end

    # GET /api/v1/private/evals/runs
    # List eval run results for the caller's tenant.
    def list_runs(params = {})
      @http.get("/api/v1/private/evals/runs", params)
    end

    # GET /api/v1/private/evals/summary
    # Get aggregate summary of eval results.
    def summary
      @http.get("/api/v1/private/evals/summary")
    end

    # GET /api/v1/private/evals/runs/:runId
    # Get detailed results for a specific eval run.
    # @param run_id [String]
    def run_detail(run_id)
      @http.get("/api/v1/private/evals/runs/#{enc(run_id)}")
    end

    # GET /api/v1/private/evals/compare
    # Compare results across multiple eval runs.
    # @param params [Hash] e.g. { "runIds" => "id1,id2" }
    def compare(params = {})
      @http.get("/api/v1/private/evals/compare", params)
    end

    # GET /api/v1/private/evals/memory-graph
    # Get memory graph visualization data for eval runs.
    def memory_graph(params = {})
      @http.get("/api/v1/private/evals/memory-graph", params)
    end

    # GET /api/v1/private/evals/runs/:runId/memory
    # Get memory usage for a specific eval run.
    # @param run_id [String]
    def run_memory(run_id)
      @http.get("/api/v1/private/evals/runs/#{enc(run_id)}/memory")
    end

    # GET /api/v1/private/evals/crew-memory
    # Get agent crew memory for a pipeline.
    # @param pipeline [String]
    # @param project [String]
    def crew_memory(pipeline:, project:)
      @http.get("/api/v1/private/evals/crew-memory",
        "pipeline" => pipeline,
        "project"  => project)
    end

    # GET /api/v1/private/evals/benchmarks
    # Get benchmark scores across eval runs.
    def benchmarks(params = {})
      @http.get("/api/v1/private/evals/benchmarks", params)
    end

    private

    def enc(s)
      URI.encode_www_form_component(s.to_s)
    end
  end
end

File: packages/sdk-python/src/melaya/evals.py
"""Evals API — pipeline evaluation runs, summaries, and benchmarks.

Maps to ``/api/v1/private/evals/*``.

Example
-------
>>> from melaya import Melaya
>>> m = Melaya(api_key="mk_...")
>>> runs = m.evals.list_runs()
>>> summary = m.evals.summary()
>>> detail = m.evals.run_detail(runs[0]["id"])
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .platform_types import JsonDict


class EvalsAPI:
    def __init__(self, request: Any) -> None:
        self._request = request

    def list_runs(self) -> List[JsonDict]:
        """List eval run results for the caller's tenant."""
        return self._request("GET", "/api/v1/private/evals/runs")

    def summary(self) -> JsonDict:
        """Get aggregate summary of eval results."""
        return self._request("GET", "/api/v1/private/evals/summary")

    def run_detail(self, run_id: str) -> JsonDict:
        """Get detailed results for a specific eval run."""
        return self._request("GET", f"/api/v1/private/evals/runs/{run_id}")

    def compare(self, *, run_ids: Optional[List[str]] = None, **kwargs: Any) -> JsonDict:
        """Compare results across multiple eval runs."""
        params: Dict[str, Any] = {**kwargs}
        if run_ids is not None:
            params["runIds"] = run_ids
        return self._request("GET", "/api/v1/private/evals/compare", params=params if params else None)

    def memory_graph(self) -> JsonDict:
        """Get memory graph visualization data for eval runs."""
        return self._request("GET", "/api/v1/private/evals/memory-graph")

    def run_memory(self, run_id: str) -> JsonDict:
        """Get memory usage for a specific eval run."""
        return self._request("GET", f"/api/v1/private/evals/runs/{run_id}/memory")

    def crew_memory(self, *, pipeline: str, project: str) -> JsonDict:
        """Get agent crew memory for a pipeline.

        Args:
            pipeline: The pipeline name.
            project: The project name.
        """
        return self._request("GET", "/api/v1/private/evals/crew-memory",
                             params={"pipeline": pipeline, "project": project})

    def benchmarks(self) -> JsonDict:
        """Get benchmark scores across eval runs."""
        return self._request("GET", "/api/v1/private/evals/benchmarks")
